festim/problem.py

```diff
@@ -0,0 +1,164 @@
+import dolfinx
+from dolfinx import fem
+from dolfinx.nls.petsc import NewtonSolver
+import basix
+import ufl
+from mpi4py import MPI
+import numpy as np
+import tqdm.autonotebook
+import festim as F
+
+
+class Problem:
+    def __init__(
+        self,
+        mesh=None,
+        sources=None,
+        exports=None,
+        subdomains=None,
+        boundary_conditions=None,
+        settings=None,
+    ) -> None:
+        self.mesh = mesh
+        # for arguments to initialise as empty list
+        # if arg not None, assign arg, else assign empty list
+        self.subdomains = subdomains or []
+        self.boundary_conditions = boundary_conditions or []
+        self.sources = sources or []
+        self.exports = exports or []
+        self.settings = settings
+
+        self.dx = None
+        self.ds = None
+        self.function_space = None
+        self.facet_meshtags = None
+        self.volume_meshtags = None
+        self.formulation = None
+        self.bc_forms = []
+
+    @property
+    def volume_subdomains(self):
+        return [s for s in self.subdomains if isinstance(s, F.VolumeSubdomain)]
+
+    @property
+    def surface_subdomains(self):
+        return [s for s in self.subdomains if isinstance(s, F.SurfaceSubdomain)]
+
+    def define_function_spaces(self):
+        """Creates the function space of the model, creates a mixed element if
+        model is multispecies. Creates the main solution and previous solution
+        function u and u_n. Create global DG function spaces of degree 0 and 1
+        for the global diffusion coefficient"""
+
+        degree = 1
+        element_CG = basix.ufl.element(
+            basix.ElementFamily.P,
+            self.mesh.mesh.basix_cell(),
+            degree,
+            basix.LagrangeVariant.equispaced,
+        )
+
+        return element_CG
+
+    def define_meshtags_and_measures(self):
+        """Defines the facet and volume meshtags of the model which are used
+        to define the measures fo the model, dx and ds"""
+
+        if isinstance(self.mesh, F.MeshFromXDMF):
+            self.facet_meshtags = self.mesh.define_surface_meshtags()
+            self.volume_meshtags = self.mesh.define_volume_meshtags()
```

Comment:
I notice here a naming inconsistency (not related to this PR) between `surface_meshtags` and `facet_mesthags`

Reply:
added a TODO for this